x-pack/plugins/security_solution/public/detection_engine/rule_exceptions/pages/exception_list_details/list_with_search/use_list_with_search.tsx

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useEffect, useMemo, useState } from 'react';
+
+import type { ExceptionListSchema } from '@kbn/securitysolution-io-ts-list-types';
+
+import type { GetExceptionItemProps } from '../../../components/list_common';
+import { ViewerStatus } from '../../../components/list_common';
+import {
+  prepareFetchExceptionItemsParams,
+  fetchListExceptionItems,
+  getExceptionItemsReferences,
+} from '../list_api';
+import * as i18n from '../translations';
+import { useExceptionListDetailsContext } from '../context/exception_list_details.context';
+
+export const useListWithSearchComponent = (list: ExceptionListSchema) => {
+  const {
+    isReadOnly,
+    toasts,
+    http,
+    pagination,
+    exceptions,
+    exceptionListReferences,
+    setExceptions,
+    setPagination,
+    setExceptionListReferences,
+  } = useExceptionListDetailsContext();
+  const [viewerStatus, setViewerStatus] = useState<ViewerStatus | string>(ViewerStatus.LOADING);
+
+  const [lastUpdated, setLastUpdated] = useState<null | string | number>(null);
+
+  const handleErrorStatus = useCallback(
+    (error) => {
+      toasts?.addError(error, {
+        title: i18n.EXCEPTION_ERROR_TITLE,
+        toastMessage: i18n.EXCEPTION_ERROR_DESCRIPTION,
+      });
+      setViewerStatus(ViewerStatus.ERROR);
+    },
+    [toasts]
+  );
+  const getReferences = useCallback(async () => {
+    try {
+      const result = await getExceptionItemsReferences(list);
+      setExceptionListReferences(result);
+    } catch (error) {
+      handleErrorStatus(error);
+    }
+  }, [handleErrorStatus, list, setExceptionListReferences]);
+
+  const updateViewer = useCallback(
+    (paginationResult, dataLength, viewStatus) => {
+      setPagination(paginationResult);
+      setLastUpdated(Date.now());
+
+      if (viewStatus === ViewerStatus.EMPTY_SEARCH) setViewerStatus(!dataLength ? viewStatus : '');
+      else setViewerStatus(!dataLength ? ViewerStatus.EMPTY : '');
+    },
+    [setPagination]
+  );
+
+  const fetchItems = useCallback(
+    async (options?, viewStatus?) => {
+      try {
+        const { data, pagination: paginationResult } = await fetchListExceptionItems({
+          http,
+          ...prepareFetchExceptionItemsParams(null, list, options),
+        });
+        setExceptions(data);
+        getReferences();
+        updateViewer(paginationResult, data.length, viewStatus);
+      } catch (error) {
+        handleErrorStatus(error);
+      }
+    },
+    [http, list, setExceptions, getReferences, updateViewer, handleErrorStatus]
+  );
+
+  useEffect(() => {
+    fetchItems(null, ViewerStatus.LOADING);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```

Comment:
Does adding `fetchItems` result in numerous unwanted calls? In general, we try to avoid using that lint disabler as we've run into a number of bugs before related to disabling it.

Reply:
> This is really awesome and making these components generic will clean up a lot of the exceptions code. Now that we're past 8.5 FF, I think it'd be great to move these into a package. Any potential issues with the package, we have plenty of time to address - but also feel free not to do that in this PR if you want to keep things moving.
> 
> Thanks so much for all of this. I'm gonna pull down to play around with it, but none of my comments were anything that would block this from going in.

Thank you so much @yctercero ^_^, sure let's decide on which package